aten/src/THC/THCTensorMathReduce.cuh

```diff
@@ -16,330 +16,232 @@
 #include <thrust/system/cuda/execution_policy.h>
 #endif
 
-// Reduction operators that support `half`, unlike Thrust
-template <typename InT, typename AccT>
-struct ReduceAdd {
-  inline __device__ AccT operator()(AccT a, InT b) const {
-    return a + (AccT) b;
-  }
-};
-
-#ifdef CUDA_HALF_TENSOR
-template <>
-struct ReduceAdd<half, half> {
-  inline __device__ half operator()(half a, half b) const {
-#ifdef CUDA_HALF_INSTRUCTIONS
-    return __hadd(a, b);
-#else
-    float fa = __half2float(a);
-    float fb = __half2float(b);
-    return __float2half(fa + fb);
-#endif
-  }
-};
+/*
+Reductions that (only) operate on accumulate types. 
+*/
 
-template <>
-struct ReduceAdd<half, float> {
-  inline __device__ float operator()(float a, half b) const {
-    return a + __half2float(b);
+template <typename T>
+struct ReduceAdd {
+  inline __device__ T operator()(const T a, const T b) const {
```

Comment:
Here, for example, we can see that ReduceAdd now accepts two elements of the same type, and outputs that same type. Historically, however, it would take the input type and the accumulate type. In practice these operations all first "upconverted" the input type to the accumulate type. The updated kernels simply do this for them.